ui/app/selectors/permissions.js

```diff
@@ -0,0 +1,59 @@
+
+import { createSelector, createSelectorCreator, defaultMemoize } from 'reselect'
+import deepEqual from 'fast-deep-equal'
+import {
+  CAVEAT_NAMES,
+} from '../../../app/scripts/controllers/permissions/enums'
+
+const permissionsSelector = (state, origin) => {
+  console.log('permissionsSelector', origin)
+  return origin && state.metamask.domains && state.metamask.domains[origin]
+}
+
+// all permissions for the origin probably too expensive for deep equality check
+const accountsPermissionSelector = createSelector(
+  permissionsSelector,
+  (domain = {}) => {
+
+    console.log('accountsPermissionSelector:domain', domain)
+    return (
+      Array.isArray(domain.permissions)
+        ? domain.permissions.find(
+          perm => perm.parentCapability === 'eth_accounts'
+        )
+        : {}
+    )
+  }
+)
+
+// comparing caveats should be cheap, at least for now
+const createCaveatEqualSelector = createSelectorCreator(
+  defaultMemoize,
+  (a = {}, b = {}) => deepEqual(a.caveats, b.caveats)
```

Comment:
We should be ensuring that any updates to Redux state are immutable - a _lot_ of code assumes that to be the case. Do you suspect that caveats are being mutated?

Reply:
We're definitely not directly mutating caveats. That `deepEqual` call is just to replace the default equality operation (`===`) of Reselect per: https://github.com/reduxjs/reselect#createselectorcreatormemoize-memoizeoptions

Reply:
The default equality operation is exactly what we want though, isn't it? Why is `deepEqual` preferred in this case?

Reply:
Hm. So, here's what happens inside `rpc-cap` when we change caveats: https://github.com/MetaMask/rpc-cap/blob/master/index.ts#L606-L658

We see that a new array is instantiated whenever a single caveat is updated. So, strict object equality should work in that case.

The other case I want to cover with the `deepEqual` is false positives in case objects like arrays are reinstantiated when they're retrieved from store. That should only happen if they're being deserialized, but that should only happen on boot.

So, is it safe to say that the `deepEqual` is unnecessary?

Reply:
Ah I see - so this wasn't done as a precaution against mutation, but to avoid recomputing things when the array instance is updated but the contents are not.

Interesting :thinking: We assume throughout the codebase that if the instance is different, the state has changed. You might be right that the serialization results in a new instance each time though, resulting in unnecessary renders. I'm not sure if that happens offhand, but this feels like the wrong place to guard against it if so (e.g. we could check in the state update itself, to avoid unnecessary re-renders and computation globally in the UI). I would suggest using a plain strict equals here, and then addressing performance issues at the store level if we find that we need to.

Reply:
Agreed. I removed `createCaveatSelector` entirely and replaced it with `createSelector`.